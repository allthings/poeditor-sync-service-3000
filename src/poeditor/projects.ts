import api from './api'

export interface InterfaceProject {
  readonly id: number
  readonly isDefault: boolean
  readonly name: string
  readonly variation: string | undefined
  readonly normative: string | undefined
}

export interface InterfaceProjectFilter {
  readonly name?: string
  readonly variation?: string
  readonly normative?: string
}

export function getProjectNameAndVariation(
  projectName: string
): {
  readonly isDefault: boolean
  readonly name: string | undefined
  readonly variation: string | undefined
  readonly normative: string | undefined
} {
  const clean = projectName.toLowerCase() || ''
  const [
    ,
    name,
    ,
    /* above skipped index is 2nd regex group, used as delimiter */
    variation,
    normative,
    isDefault,
  ] = clean.match(
    /^([\w]+)([\W]*)([\w\s]*)(?:\2*)([\w\s]*?)([\s]*\(default\))*$/
  ) || [undefined, undefined, undefined, undefined, undefined, false]

  return {
    isDefault: !!isDefault,
    name: typeof name === 'string' && name.length ? name : undefined,
    normative:
      typeof normative === 'string' && normative.length ? normative : undefined,
    variation:
      typeof variation === 'string' && variation.length ? variation : undefined,
  }
}

// Wrapper for https://poeditor.com/docs/api#projects_list
export default async function listProjects(
  filters?: InterfaceProjectFilter
): Promise<ReadonlyArray<InterfaceProject>> {
  const response = await api('projects/list')

  const projects =
    response.result &&
    response.result.projects.map(({ id, name }: any) => ({
      id,
      ...getProjectNameAndVariation(name),
    }))

  if (filters) {
    return projects.filter((project: InterfaceProject) =>
      Object.keys(filters).reduce(
        (include: boolean, filter: keyof InterfaceProjectFilter) =>
          include && filters[filter] === project[filter],
        true
      )
    )
  }

  return projects
}
