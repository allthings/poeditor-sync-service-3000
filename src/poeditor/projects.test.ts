// tslint:disable no-expression-statement
import projects, { getProjectNameAndVariation } from './projects'

describe('The Poeditor Projects wrapper', () => {
  it('should return a response object', async () => {
    const result = await projects()
    expect(result).toMatchSnapshot()
  })

  it('should let me filter by project name', async () => {
    const result = await projects({ name: 'app' })
    expect(result.length).toBe(4)
    expect(result).toMatchSnapshot()
  })

  it('should let me filter by project variation', async () => {
    const result = await projects({ variation: 'residential' })
    expect(result.length).toBe(2)
    expect(result).toMatchSnapshot()
  })

  it('should let me filter by project normative', async () => {
    const result = await projects({ normative: 'formal' })
    expect(result.length).toBe(2)
    expect(result).toMatchSnapshot()
  })

  it('should let me filter by project name, variation, and normative', async () => {
    const result = await projects({
      name: 'app',
      normative: 'formal',
      variation: 'residential',
    })
    expect(result.length).toBe(1)
    expect(result).toMatchSnapshot()
  })

  it('should be able to parse name, variation, and default from POEditor project name', () => {
    expect(getProjectNameAndVariation('')).toEqual({
      isDefault: false,
      name: undefined,
      normative: undefined,
      variation: undefined,
    })

    expect(
      getProjectNameAndVariation('App - Residential - Informal (default)')
    ).toEqual({
      isDefault: true,
      name: 'app',
      normative: 'informal',
      variation: 'residential',
    })

    expect(getProjectNameAndVariation('App - Commercial - Formal')).toEqual({
      isDefault: false,
      name: 'app',
      normative: 'formal',
      variation: 'commercial',
    })

    expect(getProjectNameAndVariation('Cockpit')).toEqual({
      isDefault: false,
      name: 'cockpit',
      normative: undefined,
      variation: undefined,
    })
  })
})
