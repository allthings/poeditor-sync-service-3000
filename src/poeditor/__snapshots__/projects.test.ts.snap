// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Poeditor Projects wrapper should let me filter by project name 1`] = `
Array [
  Object {
    "id": 73515,
    "isDefault": true,
    "name": "app",
    "normative": "informal",
    "variation": "residential",
  },
  Object {
    "id": 136001,
    "isDefault": false,
    "name": "app",
    "normative": "formal",
    "variation": "commercial",
  },
  Object {
    "id": 136003,
    "isDefault": false,
    "name": "app",
    "normative": "informal",
    "variation": "commercial",
  },
  Object {
    "id": 136005,
    "isDefault": false,
    "name": "app",
    "normative": "formal",
    "variation": "residential",
  },
]
`;

exports[`The Poeditor Projects wrapper should let me filter by project name, variation, and normative 1`] = `
Array [
  Object {
    "id": 136005,
    "isDefault": false,
    "name": "app",
    "normative": "formal",
    "variation": "residential",
  },
]
`;

exports[`The Poeditor Projects wrapper should let me filter by project normative 1`] = `
Array [
  Object {
    "id": 136001,
    "isDefault": false,
    "name": "app",
    "normative": "formal",
    "variation": "commercial",
  },
  Object {
    "id": 136005,
    "isDefault": false,
    "name": "app",
    "normative": "formal",
    "variation": "residential",
  },
]
`;

exports[`The Poeditor Projects wrapper should let me filter by project variation 1`] = `
Array [
  Object {
    "id": 73515,
    "isDefault": true,
    "name": "app",
    "normative": "informal",
    "variation": "residential",
  },
  Object {
    "id": 136005,
    "isDefault": false,
    "name": "app",
    "normative": "formal",
    "variation": "residential",
  },
]
`;

exports[`The Poeditor Projects wrapper should return a response object 1`] = `
Array [
  Object {
    "id": 73515,
    "isDefault": true,
    "name": "app",
    "normative": "informal",
    "variation": "residential",
  },
  Object {
    "id": 86419,
    "isDefault": false,
    "name": "cockpit",
    "normative": undefined,
    "variation": undefined,
  },
  Object {
    "id": 136001,
    "isDefault": false,
    "name": "app",
    "normative": "formal",
    "variation": "commercial",
  },
  Object {
    "id": 136003,
    "isDefault": false,
    "name": "app",
    "normative": "informal",
    "variation": "commercial",
  },
  Object {
    "id": 136005,
    "isDefault": false,
    "name": "app",
    "normative": "formal",
    "variation": "residential",
  },
]
`;
